### PR TITLE
[NN] GatedGCNConv

### PR DESCRIPTION
**Pull request by paoxiaode**
## Description
add GatedGCNConv

GatedGCN example: https://github.com/dmlc/dgl/pull/5570
## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by frozenbugs** [comment]
@dgl-bot

**Comment by mufeili** [comment]
@dgl-bot

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+import torch
+import torch.nn.functional as F
+from torch import nn
+
+from .... import function as fn
+
+
+class GatedGCNConv(nn.Module):
+    r"""GatedGCN: Residual Gated Graph ConvNets
+        <https://arxiv.org/pdf/1711.07553v2.pdf>`
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:12]:
```suggestion
    <https://arxiv.org/pdf/1711.07553v2.pdf>`
```

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+    .. math::
+        e_{ij}^{l+1}=D^{l} h_{i}^{l}+E^{l} h_{j}^{l}+C^l e_{ij}^{l}
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:15]: Explain `D^{l}`, `h_{i}^{l}`,  `E^{l}`, and `C^l`.

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+        norm_{ij}=\Sigma_{j\in N_{i}} \sigma\left(e_{ij}^{l+1}\right)+\varepsilon
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:17]: Explain `\sigma` and `\varepsilon`

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+        \hat{e}_{ij}^{l+1}=\sigma(e_{ij}^{l+1}) \div norm_{ij}
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:19]: Is this simply dividing `\sigma(e_{ij}^{l+1})` by `norm_{ij}`? If so, consider using `\frac{}{}` or simply `/`.

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+        h_{i}^{l+1}=A^l h_{i}^{l}+\Sigma_{j \in N_{i}} \hat{e}_{ij}^{l+1} \odot B^l h_{j}^{l}
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:21]: Explain `A^l` and `B^l`.

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    ...                         activation=F.relu)
+    >>> new_node_feats, new_edge_feats = gatedGCN(graph, node_feats, edge_feats)
+    >>> new_node_feats.shape, new_edge_feats.shape
+    (torch.Size([8, 20]), torch.Size([30, 20]))
+
+    """
+
+    def __init__(
+        self,
+        input_feats,
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:66]: `in_feats` above and `input_feats` here

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+        norm_{ij}=\Sigma_{j\in N_{i}} \sigma\left(e_{ij}^{l+1}\right)+\varepsilon
+
+        \hat{e}_{ij}^{l+1}=\sigma(e_{ij}^{l+1}) \div norm_{ij}
+
+        h_{i}^{l+1}=A^l h_{i}^{l}+\Sigma_{j \in N_{i}} \hat{e}_{ij}^{l+1} \odot B^l h_{j}^{l}
+
+    Parameters
+    ----------
+    in_feats : int
+        Input feature size; i.e, the number of dimensions of :math:`\mathbf{x}_i`.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:26]: Where should `\mathbf{x}_i` come into play in the equations above?

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    edge_feats: int
+        Edge feature size; i.e., the number of dimensions of :math:\mathbf{e}_{j,i}`.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:28]: \mathbf{e}_{j,i}^{l} for any l?

Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:28]:
```suggestion
        Edge feature size; i.e., the number of dimensions of :math:`\mathbf{e}_{j,i}`.
```

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+    >>> new_node_feats.shape, new_edge_feats.shape
+    (torch.Size([8, 20]), torch.Size([30, 20]))
+
+    """
+
+    def __init__(
+        self,
+        input_feats,
+        edge_feats,
+        output_feats,
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:68]: `out_feats` above and `output_feats` here

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+        h_{i}^{l+1}=A^l h_{i}^{l}+\Sigma_{j \in N_{i}} \hat{e}_{ij}^{l+1} \odot B^l h_{j}^{l}
+
+    Parameters
+    ----------
+    in_feats : int
+        Input feature size; i.e, the number of dimensions of :math:`\mathbf{x}_i`.
+    edge_feats: int
+        Edge feature size; i.e., the number of dimensions of :math:\mathbf{e}_{j,i}`.
+    out_feats : int
+        Output feature size; i.e., the number of dimensions of :math:`h_i^{(t+1)}`.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:30]: Should "t" be "l"? Also is this the same for any l?

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+    dropout : float, optional
+        Dropout rate on node and edge feature. Defaults: ``0``.
+    batch_norm : bool
+        Whether to include batch normalization on node . Default: ``True``.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:34]: node what? node feature?

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+    ----------
+    in_feats : int
+        Input feature size; i.e, the number of dimensions of :math:`\mathbf{x}_i`.
+    edge_feats: int
+        Edge feature size; i.e., the number of dimensions of :math:\mathbf{e}_{j,i}`.
+    out_feats : int
+        Output feature size; i.e., the number of dimensions of :math:`h_i^{(t+1)}`.
+    dropout : float, optional
+        Dropout rate on node and edge feature. Defaults: ``0``.
+    batch_norm : bool
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:33]:
```suggestion
    batch_norm : bool, optional
```

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+    Parameters
+    ----------
+    in_feats : int
+        Input feature size; i.e, the number of dimensions of :math:`\mathbf{x}_i`.
+    edge_feats: int
+        Edge feature size; i.e., the number of dimensions of :math:\mathbf{e}_{j,i}`.
+    out_feats : int
+        Output feature size; i.e., the number of dimensions of :math:`h_i^{(t+1)}`.
+    dropout : float, optional
+        Dropout rate on node and edge feature. Defaults: ``0``.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:32]:
```suggestion
        Dropout rate on node and edge feature. Default: ``0``.
```
for consistency with the cases below

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+    batch_norm : bool
+        Whether to include batch normalization on node . Default: ``True``.
+    residual : bool
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:35]:
```suggestion
    residual : bool, optional
```

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+        Whether to include residual connection . Default: ``True``.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:36]:
```suggestion
        Whether to include residual connections. Default: ``True``.
```

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+    activation : callable activation function/layer or None, optional.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:37]:
```suggestion
    activation : callable activation function/layer or None, optional
```

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+        If not None, applies an activation function to the updated node features.
+        Default: ``None``.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:39]: None here and ReLU below

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+    out_feats : int
+        Output feature size; i.e., the number of dimensions of :math:`h_i^{(t+1)}`.
+    dropout : float, optional
+        Dropout rate on node and edge feature. Defaults: ``0``.
+    batch_norm : bool
+        Whether to include batch normalization on node . Default: ``True``.
+    residual : bool
+        Whether to include residual connection . Default: ``True``.
+    activation : callable activation function/layer or None, optional.
+        If not None, applies an activation function to the updated node features.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:38]:
```suggestion
        If not None, apply an activation function to the updated node features.
```

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+        Default: ``None``.
+
+    Example
+    -------
+    >>> import dgl
+    >>> import torch as th
+    >>> import torch.nn.functional as F
+    >>> from dgl.nn import GatedGCNConv
+    >>> num_nodes, num_edges = 8, 30
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:47]: add an empty line above

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+    >>> graph = dgl.rand_graph(num_nodes,num_edges)
+    >>> node_feats = th.rand((num_nodes, 20))
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:49]: It suffices to have `th.rand(num_nodes, 20)`. Same for the line below.

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+    >>> edge_feats = th.rand((num_edges, 12))
+    >>> gatedGCN = GatedGCNConv(input_feats=20,
+    ...                         edge_feats=12,
+    ...                         output_feats=20,
+    ...                         dropout=0.2,
+    ...                         batch_norm=True,
+    ...                         residual=True,
+    ...                         activation=F.relu)
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:57]: You probably need to update this if you decide to update the arguments above. Simplify this model initialization step and use default values as much as possible.

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 22]
```diff
@@ -0,0 +1,180 @@
...
+        edge_feats,
+        output_feats,
+        dropout=0,
+        batch_norm=True,
+        residual=True,
+        activation=F.relu,
+    ):
+        super(GatedGCNConv, self).__init__()
+        self._input_feats = input_feats
+        self._output_feats = output_feats
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:76]: L75 - 76 not used and may be unnecessary

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 23]
```diff
@@ -0,0 +1,180 @@
...
+        self.dropout = nn.Dropout(dropout)
+        self.batch_norm = batch_norm
+        self.residual = residual
+
+        if input_feats != output_feats or edge_feats != output_feats:
+            self.residual = False
+
+        # Linearly tranform the node features.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:84]:
```suggestion
        # Linearly transform the node features.
```

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 24]
```diff
@@ -0,0 +1,180 @@
...
+        self.A = nn.Linear(input_feats, output_feats, bias=True)
+        self.B = nn.Linear(input_feats, output_feats, bias=True)
+        self.D = nn.Linear(input_feats, output_feats, bias=True)
+        self.E = nn.Linear(input_feats, output_feats, bias=True)
+
+        # Linearly tranform the edge features.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:90]: ditto

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 25]
```diff
@@ -0,0 +1,180 @@
...
+        ----------
+        graph : DGLGraph
+            The graph.
+        feat : torch.Tensor
+            The input feature of shape :math:`(N, D_{in})` where :math:`N`
+            is the number of nodes of the graph and :math:`D_{in}` is the
+            input feature size.
+        edge_feat : torch.Tensor
+            The input edge feature of shape :math:`(E, D_{in_{edge}})`,
+            where :math:`E` is the number of edges and :math:`D_{in_{edge}}`
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:116]: Perhaps it's easier to use `D_{node}` in place of `D_{in}` above and `D_{edge}` here.

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 26]
```diff
@@ -0,0 +1,180 @@
...
+            the size of the edge features.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:117]:
```suggestion
            is the size of the edge features.
```

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 27]
```diff
@@ -0,0 +1,180 @@
...
+
+        Returns
+        -------
+        torch.Tensor
+            The output feature of shape :math:`(N, D_{out})` where :math:`D_{out}`
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:122]:
```suggestion
            The output node feature of shape :math:`(N, D_{out})` where :math:`D_{out}`
```

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 28]
```diff
@@ -0,0 +1,180 @@
...
+            The output edge feature of shape :math:`(E, D_{out})` where :math:`D_{out}`
+            is the output feature size.
+        """
+        with graph.local_scope():
+            # For residual connection
+            h_in = feat
+            e_in = edge_feat
+
+            # Linearly tranform the node features.
+            graph.ndata["h"] = feat
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:134]: not needed

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 29]
```diff
@@ -0,0 +1,180 @@
...
+        torch.Tensor
+            The output edge feature of shape :math:`(E, D_{out})` where :math:`D_{out}`
+            is the output feature size.
+        """
+        with graph.local_scope():
+            # For residual connection
+            h_in = feat
+            e_in = edge_feat
+
+            # Linearly tranform the node features.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:133]: ditto

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 30]
```diff
@@ -0,0 +1,180 @@
...
+            graph.ndata["h"] = feat
+            graph.ndata["Ah"] = self.A(feat)
+            graph.ndata["Bh"] = self.B(feat)
+            graph.ndata["Dh"] = self.D(feat)
+            graph.ndata["Eh"] = self.E(feat)
+
+            # Linearly tranform the edge features.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:140]: ditto

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 31]
```diff
@@ -0,0 +1,180 @@
...
+            graph.edata["e"] = edge_feat
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:141]: not needed

File: tests/python/pytorch/nn/test_nn.py [hunk 32]
```diff
@@ -1066,6 +1066,22 @@ def test_gated_graph_conv_one_etype(g, idtype):
...
+    g = g.astype(idtype).to(ctx)
+    gatedgcnconv = nn.GatedGCNConv(10, 10, 5)
+    feat = F.randn((g.num_nodes(), 10))
+    efeat = F.randn((g.num_edges(), 10))
+    gatedgcnconv = gatedgcnconv.to(ctx)
+
+    h, edge_h = gatedgcnconv(g, feat, efeat)
+    # current we only do shape check
+    assert h.shape == (g.number_of_dst_nodes(), 5)
+    assert edge_h.shape == (g.number_of_edges(), 5)
```
Comment (mufeili) [inline tests/python/pytorch/nn/test_nn.py:1082]: Could you move this function to a new file `tests/python/pytorch/nn/conv/test_gatedgcnconv.py`?

**Review by mufeili — COMMENTED** [review]
done a pass

**Comment by paoxiaode** [comment]
@mufeili

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 33]
```diff
@@ -0,0 +1,171 @@
...
+    .. math::
+        e_{ij}^{l+1}=D^l h_{i}^{l}+E^l h_{j}^{l}+C^l e_{ij}^{l}
+
+        norm_{ij}=\Sigma_{j\in N_{i}} \sigma\left(e_{ij}^{l+1}\right)+\varepsilon
+
+        \hat{e}_{ij}^{l+1}=\sigma(e_{ij}^{l+1}) / norm_{ij}
+
+        h_{i}^{l+1}=A^l h_{i}^{l}+\Sigma_{j \in N_{i}} \hat{e}_{ij}^{l+1} \odot B^l h_{j}^{l}
+
+    where h_{i}^{l} are node features of layer l, \Sigma is sigmoid function,
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:23]:
```suggestion
    where :math:`h_{i}^{l}` are node features of layer :math:`l`, :math:`\Sigma` is sigmoid function,
```

Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:23]: missing the explanation for `e_{ij}^{l}`

**Comment by mufeili** [comment]
I cannot find the description of the model with edge features in the original paper. Where did it come from?

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 34]
```diff
@@ -0,0 +1,171 @@
...
+    \varepsilon is miniterm, A^l, B^l, C^l, D^l, E^l are linear layers.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:24]:
```suggestion
    :math:`\varepsilon` is miniterm, :math:`A^l, B^l, C^l, D^l, E^l` are linear layers.
```

Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:24]: Where is \varepsilon defined in the original paper?

File: tests/python/pytorch/nn/conv/test_gatedgcnconv.py [hunk 35]
```diff
@@ -0,0 +1,30 @@
+import io
+import pickle
+from copy import deepcopy
```
Comment (mufeili) [inline tests/python/pytorch/nn/conv/test_gatedgcnconv.py:3]: L2-3 not used

File: tests/python/pytorch/nn/conv/test_gatedgcnconv.py [hunk 36]
```diff
@@ -0,0 +1,30 @@
...
+
+import backend as F
+
+import dgl
```
Comment (mufeili) [inline tests/python/pytorch/nn/conv/test_gatedgcnconv.py:7]: L7 not used

File: tests/python/pytorch/nn/conv/test_gatedgcnconv.py [hunk 37]
```diff
@@ -0,0 +1,30 @@
...
+import dgl.nn.pytorch as nn
+import pytest
+import torch
```
Comment (mufeili) [inline tests/python/pytorch/nn/conv/test_gatedgcnconv.py:10]: L10 not used

**Comment by mufeili** [comment]
@dgl-bot

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py (hunk 34 above)
Comment (paoxiaode) [reply python/dgl/nn/pytorch/conv/gatedgcnconv.py]: add missing paper: Benchmarking Graph Neural Networks

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 38]
```diff
@@ -0,0 +1,175 @@
...
+# pylint: disable= no-member, arguments-differ, invalid-name, cell-var-from-loop
+import torch
+import torch.nn.functional as F
+from torch import nn
+
+from .... import function as fn
+
+
+class GatedGCNConv(nn.Module):
+    r"""GatedGCN: Residual Gated Graph ConvNets
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:11]:
```
Gated graph convolutional layer from `Benchmarking Graph Neural Networks <https://arxiv.org/abs/2003.00982>`__
```
is sufficient

Comment (mufeili) [reply python/dgl/nn/pytorch/conv/gatedgcnconv.py:11]: ``__ is used for hyperlink in doc rendering.

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 39]
```diff
@@ -0,0 +1,175 @@
...
+
+        norm_{ij}=\Sigma_{j\in N_{i}} \sigma\left(e_{ij}^{l+1}\right)+\varepsilon
+
+        \hat{e}_{ij}^{l+1}=\sigma(e_{ij}^{l+1}) / norm_{ij}
+
+        h_{i}^{l+1}=A^l h_{i}^{l}+\Sigma_{j \in N_{i}} \hat{e}_{ij}^{l+1} \odot B^l h_{j}^{l}
+
+    where :math:`h_{i}^{l}` is node :math:`i` feature of layer :math:`l`,
+    :math:`e_{ij}^{l}` is edge :math:`ij` feature of layer :math:`l`,
+    :math:`\Sigma` is sigmoid function, :math:`\varepsilon` is miniterm,
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:27]: `\Sigma` is the summation symbol, and `\sigma` is widely used for activation functions like sigmoid.

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 40]
```diff
@@ -0,0 +1,173 @@
...
+
+        norm_{ij}=\Sigma_{j\in N_{i}} \sigma\left(e_{ij}^{l+1}\right)+\varepsilon
+
+        \hat{e}_{ij}^{l+1}=\sigma(e_{ij}^{l+1}) / norm_{ij}
+
+        h_{i}^{l+1}=A^l h_{i}^{l}+\Sigma_{j \in N_{i}} \hat{e}_{ij}^{l+1} \odot B^l h_{j}^{l}
+
+    where :math:`h_{i}^{l}` is node :math:`i` feature of layer :math:`l`,
+    :math:`e_{ij}^{l}` is edge :math:`ij` feature of layer :math:`l`,
+    :math:`\sigma` is sigmoid function, :math:`\varepsilon` is miniterm,
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:25]:
```suggestion
    :math:`\sigma` is sigmoid function, :math:`\varepsilon` is a small fixed constant for numerical stability,
```

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 41]
```diff
@@ -0,0 +1,173 @@
...
+
+
+class GatedGCNConv(nn.Module):
+    r"""Gated graph convolutional layer from `Benchmarking Graph Neural Networks
+    <https://arxiv.org/abs/2003.00982>`__
+
+    .. math::
+        e_{ij}^{l+1}=D^l h_{i}^{l}+E^l h_{j}^{l}+C^l e_{ij}^{l}
+
+        norm_{ij}=\Sigma_{j\in N_{i}} \sigma\left(e_{ij}^{l+1}\right)+\varepsilon
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:17]: In the original paper, it should be

```
norm_{ij}=\Sigma_{j\in N_{i}} \sigma\left(e_{ij}^{l+1}\right)+\varepsilon+e_{ij}^{l}

\hat{e}_{ij}^{l+1}=\sigma(e_{ij}^{l+1} + e_{ij}^{l}) / norm_{ij}
```
right?

Comment (mufeili) [reply python/dgl/nn/pytorch/conv/gatedgcnconv.py:17]: Also you interchange the definition of \hat{e} and e, which can be confusing. It's better to be consistent with the paper notations.

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 42]
```diff
@@ -0,0 +1,173 @@
...
+
+        \hat{e}_{ij}^{l+1}=\sigma(e_{ij}^{l+1}) / norm_{ij}
+
+        h_{i}^{l+1}=A^l h_{i}^{l}+\Sigma_{j \in N_{i}} \hat{e}_{ij}^{l+1} \odot B^l h_{j}^{l}
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:21]: ditto

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 43]
```diff
@@ -0,0 +1,173 @@
...
+            graph.ndata["Bh"] = self.B(feat)
+            graph.ndata["Dh"] = self.D(feat)
+            graph.ndata["Eh"] = self.E(feat)
+            graph.edata["Ce"] = self.C(edge_feat)
+
+            graph.apply_edges(fn.u_add_v("Dh", "Eh", "DEh"))
+
+            # Get edge feature
+            graph.edata["e"] = graph.edata["DEh"] + graph.edata["Ce"]
+            graph.edata["sigma"] = torch.sigmoid(graph.edata["e"])
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:141]: From the paper, the following operations are missing between L140 and 141 in the paper:

- batch normalization
- ReLU
- residual connection

File: python/dgl/nn/pytorch/conv/gatedgcnconv.py [hunk 44]
```diff
@@ -0,0 +1,173 @@
...
+
+            # Non-linear activation.
+            if self.activation:
+                feat = self.activation(feat)
+                edge_feat = self.activation(edge_feat)
+
+            # Residual connection.
+            if self.residual:
+                feat = h_in + feat
+                edge_feat = e_in + edge_feat
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatedgcnconv.py:168]: From the paper, batch normalization, activation, residual for edge features should happen before.

**Review by mufeili — APPROVED** [review]
I'm good other than some final minor comments.

**Comment by mufeili** [comment]
@dgl-bot

**Pull request merged**